internal/graph/singleflight_resolver.go

```diff
@@ -2,10 +2,12 @@ package graph
 
 import (
 	"context"
+	"sync/atomic"
 
 	"go.opentelemetry.io/otel/attribute"
 	"go.uber.org/zap"
-	"golang.org/x/sync/singleflight"
+
+	"resenje.org/singleflight"
```

Comment:
This change was actually critical to fix the broken tests across the board. In [resenje.org/singleflight#Group.Do](https://pkg.go.dev/resenje.org/singleflight#Group.Do) 

> if there are multiple callers, context passed to one caller does not affect the execution and returned values of others except if the function result is dependent on the context values.

This ☝️  was the exact problem manifesting itself. We'd de-duplicate with different contexts passed in (see [this test](https://github.com/openfga/openfga/pull/1335/files#diff-7031b8db92a3ea494a5f79c70542afa86c66750a16601d78a4cdedf3bddc4536R154)) at the top-level and yet one would return a context cancellation error. 

The regular [x/sync/singleflight#Group.Do](https://pkg.go.dev/golang.org/x/sync/singleflight#Group.Do) creates a closure over the context that wins the singleflight race, and thus causes some perplexing context cancellation errors in the de-duplicated paths.

The `resenje.org/singleflight` package avoids this issue with better support for context for singleflight invocations. Notice that we pass in `ctx` to `Group.Do(ctx, ...)` but the function that gets executed once for each of the shared invocations gets it's own unique `innerCtx context.Context`.

```
Do(ctx, ..., func(innerCtx context.Context) ... {
    _ = innerCtx // we use innerCtx to avoid shared closure over `ctx`
})
```

Reply:
In fact, I have found that what @willvedd and I were exploring in https://github.com/openfga/openfga/pull/1335#issuecomment-1942938328 was also a bit misleading. I have pretty high confidence now that we were wrong in our theory.

The root cause still boils down to context propagation across the inflight singleflight resolvers. All of our hypothesis' around tracking locked subproblems didn't work (tests failed) when I ran large iterations of our tests. They led to more stable tests, but I still saw intermittent test failures quite frequently.

The changes as they are now are the latest most changes that fix the problem in a stable way. I can run our test suites for ListObjects and Check with a large number of iterations and they still pass. The changes also make the most sense to me, becase prior to using the `resenje.org/singlelfight` package we were always going to have some problems with independent context overlapping on de-duplicated subproblems. I wish the standard `x/sync/singleflight` package has `DoWithContext` that operated like this alternative package does.